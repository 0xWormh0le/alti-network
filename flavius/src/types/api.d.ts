type QueryParamValue = string | number | string[] | number[] | null | undefined | boolean
type QueryParams = { [k: string]: QueryParamValue }

type ApiClientReturn<TState extends any, TCallerArgs extends any> = [
  Maybe<TState>,
  Maybe<{ error: string }>,
  boolean,
  ApiCaller<TCallerArgs>
]

type ApiErrorCallback = (err: any) => void
type ApiSuccessCallback = (response: any) => void

interface ApiClientInit {
  defaultPageSize?: number
  pageCountCacheTtl?: number
  handleError?: ApiErrorCallback
  handleSuccess?: ApiSuccessCallback
  immediateTrigger?: boolean
}

interface ApiCaller<T> {
  call: (args?: T) => void
}

interface PaginationArgs {
  pageSize: number
  pageCountCacheTtl: number
  pageCountLastUpdated: number
}

interface MetaArgs<T, TParams = unknown> {
  handleError?: (err: Error) => void
  handleSuccess?: (rs: T) => void
  params?: (QueryParams & TParams) | undefined
  preventFirstCall?: boolean
  triggeringKeys?: string[]
}

interface InternalMetaArgs<T, TParams = unknown> extends MetaArgs<T, TParams> {
  trigger: boolean
  resetTrigger: () => void
  handleError: (err: Error) => void
  handleSuccess: (rs: T) => void
  params: (QueryParams & TParams) | undefined
}
interface ApiCallArgs<T, TArgs> {
  requestDetails?: TArgs
  meta?: MetaArgs<T>
}

type WithPagination<T> = T & {
  pageNumber: number
  pageSize: number
}
