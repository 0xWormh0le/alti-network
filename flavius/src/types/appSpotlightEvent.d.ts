// Display interfaces for each kind of event types are represented below

// event type: people
interface DisplayAppSpotlightPeople {
  avatar: string
  name: string
  email: string
}

// event type: downloads
interface DisplayAppSpotlightEventDownloads {
  view: IFile
  file: IFile
  owner: IPerson
  datetime: number
}

interface BaseAppSpotlightEvent {
  targetPeople: IPerson[]
  files: IFile[]
  eventName: string
  eventDescription: string
  severity: SeverityRange
  datetime: number
}

interface AppSpotlightEvent extends BaseAppSpotlightEvent {
  eventId: string
}

// event type: risks
interface DisplayAppSpotlightRisk extends BaseRisk {
  view: FileResponseWithRisk
  file: FileCellWithRisk
  description: DisplayRiskDescription
  creator: IPerson
  owner: FileOwner
  action: RiskAction
}
