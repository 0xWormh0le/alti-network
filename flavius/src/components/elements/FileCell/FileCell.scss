@import 'styles/_shared.scss';

$ICON_SIZE: 1.2em;

.FileCell {
  &__icons {
    display: flex;
    margin-top: 4px;
  }

  &__spacer-top-5 {
    padding-top: 5px;
  }

  &__singleline {
    @include flex(row, flex-start, flex-start);
    & > a > div:nth-child(1) {
      display: flex;
    }
  }
  &__multiline {
    @include flex(row, flex-start, flex-start);
  }

  min-width: 150px;
  max-width: 200px;

  @media (min-width: 1200px) {
    max-width: 300px;
  }

  &__link {
    @include brand-link();
    overflow: hidden;
    text-overflow: ellipsis;
  }

  & > div {
    line-height: 1em;
  }
}
