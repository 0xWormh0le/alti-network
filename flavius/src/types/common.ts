export enum UsageKind {
  personal = 'personal',
  work = 'work'
}

export enum UserKind {
  user = 'user',
  person = 'person',
  serviceAccount = 'serviceAccount'
}

export enum AccessLevel {
  member = 'member',
  manager = 'manager',
  owner = 'owner',
  admin = 'admin'
}

export enum RiskCategoryType {
  SHARING_RISKS = 'Sharings',
  RELATIONSHIP_RISKS = 'Relationships',
  ACTIVITY_RISKS = 'Actions',
  INFORMATIONAL_RISKS = 'Informatives'
}

export enum FileActivityType {
  ACCESS_GRANTED = 'ACCESS_GRANTED',
  CHANGE_FOLDER_PERMISSION = 'CHANGE_FOLDER_PERMISSION',
  CONTENT_ACCESS = 'CONTENT_ACCESS',
  DOWNLOAD = 'DOWNLOAD',
  FILE_WATERMARKED_DOWNLOAD = 'FILE_WATERMARKED_DOWNLOAD',
  ITEM_DOWNLOAD = 'ITEM_DOWNLOAD',
  ITEM_PREVIEW = 'ITEM_PREVIEW',
  ITEM_SHARED = 'ITEM_SHARED',
  ITEM_SHARED_CREATE = 'ITEM_SHARED_CREATE',
  ITEM_SHARED_UPDATE = 'ITEM_SHARED_UPDATE',
  ITEM_SYNC = 'ITEM_SYNC',
  SHARE = 'SHARE'
}

export enum PermissionRole {
  own = 'own',
  read = 'read',
  write = 'write'
}

export enum PermissionShared {
  external = 'external',
  internal = 'internal',
  none = 'none'
}

export enum PermissionType {
  anyone = 'anyone',
  domain = 'domain',
  user = 'user'
}

export enum PermissionItemStatus {
  active = 'ACTIVE',
  pending = 'PENDING',
  removed = 'REMOVED',
  cannotBeRemoved = 'CANNOT_BE_REMOVED'
}

export enum SensitiveContentStatus {
  DETECTION_PENDING = 'pending',
  FOUND_SENSITIVE_CONTENT = 'yes',
  FOUND_NO_SENSITIVE_CONTENT = 'no',
  FAILED_DETECTION = 'failed'
}

export enum FileGridNavType {
  TIMELINE = 'timeline',
  INTERNAL = 'internal',
  EXTERNAL = 'external',
  SENSITIVE = 'sensitive',
  FILES_IN_FOLDER = 'filesInFolder',
  DETAILS = 'details'
}
