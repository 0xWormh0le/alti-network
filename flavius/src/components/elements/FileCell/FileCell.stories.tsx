import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import FileCell, { FileCellProps } from './FileCell'
import { BrowserRouter } from 'react-router-dom'
import { fileMock } from 'test/mocks'

export default {
  title: 'Elements/FileCell',
  component: FileCell
} as Meta

const Template: Story<FileCellProps> = (args) => (
  <BrowserRouter>
    <FileCell {...args} />
  </BrowserRouter>
)

export const Default = Template.bind({})
Default.args = {
  fileResponse: fileMock
}

export const WithFileIcon = Template.bind({})
WithFileIcon.args = {
  fileResponse: fileMock
}
