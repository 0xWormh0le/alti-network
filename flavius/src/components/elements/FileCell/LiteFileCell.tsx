import Tooltip from 'components/widgets/Tooltip'
import React from 'react'
import { Link } from 'react-router-dom'
import { ellipsify, modalBasePath } from 'util/helpers'
import FileIcon from '../FileIcon'
import { MAX_LENGTH } from './FileCell'

interface LiteFileCellProps {
  file: LiteFile
  platformId: string
}

const LiteFileCell: React.FC<LiteFileCellProps> = ({ file, platformId }) => {
  return (
    <div className='FileCell FileCell__singleline'>
      <div className='FileCell__icons'>
        <FileIcon mimeType={file.mimeType} />
      </div>
      <Link
        to={`${modalBasePath()}/file/${encodeURIComponent(file.fileId)}?owner=${
          file.createdBy?.primaryEmail?.address
        }&platformId=${platformId}`}
        className='FileCell__link'>
        {/* 
          TODO: Replace this link with the proper file spotlight link when it's ready
        */}
        <div>
          <Tooltip text={file.fileName}>
            <span>{ellipsify(file.fileName, MAX_LENGTH)}</span>
          </Tooltip>
        </div>
      </Link>
    </div>
  )
}

export default LiteFileCell
