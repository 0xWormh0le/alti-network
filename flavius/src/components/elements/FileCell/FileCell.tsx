import React, { useMemo } from 'react'
import { Link } from 'react-router-dom'
import { upperFirst } from 'lodash'
import AppIndicator from '../AppIndicator'
import SensitiveFileCell from 'components/elements/SensitiveFileCell'
import {
  ellipsify,
  modalBasePath,
  searchWithoutModalParams,
  hasSensitiveContentSimple,
  getFolderUrl,
  formatSingleFileUrl
} from 'util/helpers'
import FileIcon from 'components/elements/FileIcon'
import Tooltip from 'components/widgets/Tooltip'
import { UI_STRINGS } from 'util/ui-strings'
import { renderAttributeIfDev } from 'util/helpers'
import { SensitiveContentStatus } from 'types/common'
import './FileCell.scss'

export const MAX_LENGTH = 28
export const MAX_LENGTH_NOT_SENSITIVE = 60

export interface FileCellProps {
  fileResponse: (IFile & FileCellWithoutRisk) | FileCellWithRisk
  sensitiveContentDetected?: SensitiveContentStatus
}

const formatMultipleFilesUrl = (displayRiskDescription: FileCellWithRisk) => {
  const riskId = displayRiskDescription.riskId
  const riskTypeId = displayRiskDescription.riskTypeId
  const pluginId = displayRiskDescription.pluginId || ''
  const pluginName = displayRiskDescription.pluginName || upperFirst(UI_STRINGS.FILE.UNKNOWN)
  const personId = displayRiskDescription.personId?.replace(/\?(.*)/, '')
  const platformId = displayRiskDescription.platformId
  const modalParams = searchWithoutModalParams()
  let linkUrl = `${modalBasePath()}/files/${riskId}${modalParams}`

  if (riskId && typeof riskId !== undefined && riskId !== 'undefined') {
    const params = []
    params.push(`riskTypeId=${riskTypeId}`)
    if (pluginId) {
      params.push(`pluginId=${pluginId}`)
    }
    if (pluginName) {
      params.push(`pluginName=${pluginName}`)
    }
    if (personId) {
      params.push(`owner=${personId}`)
    }
    if (riskTypeId && typeof riskTypeId !== undefined && riskTypeId !== undefined) {
      params.push(`riskTypeId=${riskTypeId}`)
    }
    if (platformId) {
      params.push(`files-platformIds=["${platformId}"]`) // for GET /files endpoint, we will pass platformIds and prefix it with files
    }
    if (params && params.length && riskTypeId) {
      linkUrl += (modalParams ? '&' : '?') + params.join('&')
    }
  }
  return linkUrl
}

const FileCell: React.FC<FileCellProps> = ({ fileResponse, sensitiveContentDetected }) => {
  const file = fileResponse

  const { fileName, app, mimeType, iconUrl, platformId, webLink, fileId, createdBy } = file

  const fileCount = file.fileCount || 0

  const owner = createdBy?.primaryEmail?.address

  const isSensitiveContent = hasSensitiveContentSimple(sensitiveContentDetected)

  const linkUrl = useMemo(() => {
    if (fileCount > 1) {
      // There is more than one file related, link to the list of files associated with the risk
      return formatMultipleFilesUrl(file as FileCellWithRisk)
    } else {
      // Link to the file page using the `fileId`
      const isFolder = mimeType === 'folder'

      if (isFolder) {
        return getFolderUrl(fileId, platformId) || '#'
      } else {
        return formatSingleFileUrl(owner, fileId, platformId, isSensitiveContent)
      }
    }
  }, [file, fileCount, fileId, isSensitiveContent, mimeType, owner, platformId])

  const cellText = useMemo(() => {
    if (fileCount > 1) {
      return `${fileCount} files`
    } else {
      const maxLength = isSensitiveContent ? MAX_LENGTH : MAX_LENGTH_NOT_SENSITIVE
      return ellipsify(fileName, maxLength)
    }
  }, [fileCount, fileName, isSensitiveContent])

  const cssClassName = isSensitiveContent ? 'FileCell FileCell__multiline' : 'FileCell FileCell__singleline'
  const showIcon = !fileCount || fileCount === 1

  return (
    <React.Fragment>
      <div className={cssClassName}>
        <div className='FileCell__icons'>
          <AppIndicator value={app || ''} fileId={fileId} platformId={platformId} webLink={webLink} />
          <div style={{ marginRight: '0.3em' }} />
          {showIcon && <FileIcon mimeType={mimeType} iconUrl={iconUrl || ''} />}
        </div>
        <Link to={linkUrl} className='FileCell__link' {...renderAttributeIfDev({ 'data-testid': 'file-cell-link' })}>
          <div>
            {fileCount === 1 && cellText && cellText.match(/â€¦/gi) ? (
              <Tooltip text={fileName}>
                <span>{cellText}</span>
              </Tooltip>
            ) : (
              <div>{cellText}</div>
            )}
            {isSensitiveContent && file.riskId && (
              <div className='FileCell__spacer-top-5'>
                <SensitiveFileCell
                  sensitiveContentDetected={sensitiveContentDetected}
                  sensitiveFile={{
                    sensitivePhrases: file.sensitivePhrases,
                    fileCount,
                    riskId: file.riskId,
                    riskTypeId: file.riskTypeId
                  }}
                />
              </div>
            )}
          </div>
        </Link>
      </div>
    </React.Fragment>
  )
}

export default FileCell
