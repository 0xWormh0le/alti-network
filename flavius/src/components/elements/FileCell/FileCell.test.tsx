import React from 'react'
import { renderWithRouter } from 'test/support/helpers'
import { UserKind } from 'types/common'
import FileCell, { MAX_LENGTH_NOT_SENSITIVE } from './FileCell'

it('renders correctly', () => {
  const fileCellValue: FileCellWithRisk = {
    fileName: 'Test Filename',
    fileCount: 1,
    riskId: 'risk1',
    fileId: 'fileA',
    mimeType: 'mimeType',
    platformId: 'gsuite',
    riskTypeId: 1020,
    pluginId: 'pluginId',
    pluginName: 'pluginName',
    app: 'app',
    email: 'email',
    iconUrl: 'iconUrl',
    personId: 'personId',
    text: 'text',
    sensitivePhrases: {
      ccNum: 3,
      sensitiveKeywords: [],
      sensitiveKeywordsFileCount: undefined,
      ssn: 1
    } as SensitivePhraseSingleFile,
    webLink: 'webLink',
    createdBy: {
      accessLevel: 0,
      projectId: 'projectId',
      userKind: UserKind.person,
      accessCount: 0,
      altnetId: 'altnetId',
      anonymous: false,
      primaryEmail: {
        accessCount: 0,
        kind: 0,
        lastDeletedPermissions: 0,
        primary: true,
        riskCount: 1,
        address: 'hello@world.com'
      }
    }
  }
  const { container } = renderWithRouter(<FileCell fileResponse={fileCellValue} />)
  expect(container).toMatchSnapshot()
})

it('links to an individual File page when the risk is related to one file', () => {
  // mocking the `file` object that is built after the data is retrieved
  const fileCellValue: FileCellWithRisk = {
    fileName: 'Test Filename',
    fileCount: 1,
    riskId: 'risk1',
    fileId: 'fileA',
    mimeType: 'mimeType',
    sensitivePhrases: {
      ccNum: 3,
      sensitiveKeywords: [],
      sensitiveKeywordsFileCount: undefined,
      ssn: 1
    } as SensitivePhraseSingleFile,
    webLink: 'webLink',
    app: 'app',
    email: 'email',
    iconUrl: 'iconUrl',
    personId: 'personId',
    pluginId: 'pluginId',
    pluginName: 'pluginName',
    riskTypeId: 1020,
    text: 'text',
    createdBy: {
      accessLevel: 0,
      projectId: 'projectId',
      userKind: UserKind.person,
      accessCount: 0,
      altnetId: 'altnetId',
      anonymous: false,
      primaryEmail: {
        accessCount: 0,
        kind: 0,
        lastDeletedPermissions: 0,
        primary: true,
        riskCount: 1,
        address: 'hello@world.com'
      }
    },
    platformId: 'gsuite'
  }

  const { getByTestId } = renderWithRouter(<FileCell fileResponse={fileCellValue} />)

  const linkElement = getByTestId('file-cell-link')
  expect(linkElement).toHaveAttribute(
    'href',
    '/dashboard/file/fileA?owner=hello@world.com&selection=sensitive&platformId=gsuite&sensitive=true'
  )
})

it('links to a list of Files when the risk is related to many files', () => {
  const fileCellValue: FileCellWithRisk = {
    fileName: 'Test Filename',
    fileCount: 4,
    riskId: 'risk2',
    platformId: 'gsuite',
    fileId: 'fileId',
    mimeType: 'mimeType',
    sensitivePhrases: {
      ccNumFileCount: 3,
      sensitiveKeywords: undefined,
      sensitiveKeywordsFileCount: 4,
      ssnFileCount: 4
    } as SensitivePhraseMultiFile,
    webLink: 'webLink',
    app: 'app',
    email: 'email',
    iconUrl: 'iconUrl',
    personId: 'personId',
    pluginId: 'pluginId',
    pluginName: 'pluginName',
    riskTypeId: 1020,
    text: 'text',
    createdBy: {
      accessLevel: 0,
      projectId: 'projectId',
      userKind: UserKind.person,
      accessCount: 0,
      altnetId: 'altnetId',
      anonymous: false,
      primaryEmail: {
        accessCount: 0,
        kind: 0,
        lastDeletedPermissions: 0,
        primary: true,
        riskCount: 1,
        address: 'hello@world.com'
      }
    }
  }

  const { getByTestId } = renderWithRouter(<FileCell fileResponse={fileCellValue} />)

  const linkElement = getByTestId('file-cell-link')
  expect(linkElement).toHaveAttribute(
    'href',
    '/dashboard/files/risk2?riskTypeId=1020&pluginId=pluginId&pluginName=pluginName&owner=personId&riskTypeId=1020&files-platformIds=["gsuite"]'
  )
})

it('shows file count when the risk is related to many files', () => {
  const fileCellValue: any = {
    fileName: '12345678901234567890123456789012345678901',
    fileCount: 4,
    riskId: 'risk2',
    platformId: 'gsuite',
    createdBy: {
      primaryEmail: {
        address: 'hello@world.com'
      }
    }
  }

  const { getByTestId } = renderWithRouter(<FileCell fileResponse={fileCellValue} />)

  const linkElement = getByTestId('file-cell-link')
  expect(linkElement).toHaveTextContent('4 files')
})

it(`truncates file name when the length is greater than ${MAX_LENGTH_NOT_SENSITIVE}`, () => {
  // mocking the `file` object that is built after the data is retrieved
  const fileCellValue: any = {
    fileName: 'a'.repeat(MAX_LENGTH_NOT_SENSITIVE + 1),
    riskId: 'risk2',
    platformId: 'gsuite',
    createdBy: {
      primaryEmail: {
        address: 'hello@world.com'
      }
    }
  }

  const { getByTestId } = renderWithRouter(<FileCell fileResponse={fileCellValue} />)

  const linkElement = getByTestId('file-cell-link')
  expect(linkElement).toHaveTextContent('â€¦')
})
