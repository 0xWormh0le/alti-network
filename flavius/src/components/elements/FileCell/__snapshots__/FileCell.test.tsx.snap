// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    class="FileCell FileCell__multiline"
  >
    <div
      class="FileCell__icons"
    >
      <span
        class="AppIndicator"
      >
        <div
          class="platform-container"
        >
          <a
            class="trigger"
            href="webLink"
            rel="noopener noreferrer"
            target="_blank"
          >
            <div>
              <div>
                <div
                  class="platform-icon-container"
                  data-testid="platform-icon"
                >
                  <img
                    alt="Google Workspace"
                    class="platform-icon"
                    src="google-logo.svg"
                    title="Google Workspace"
                  />
                </div>
              </div>
            </div>
          </a>
        </div>
      </span>
      <div
        style="margin-right: 0.3em;"
      />
      <div
        class="file_icon_container"
      >
        <div
          class="trigger"
        >
          <img
            alt="mimeType"
            class="file_icon"
            src="iconUrl"
          />
        </div>
      </div>
    </div>
    <a
      class="FileCell__link"
      data-testid="file-cell-link"
      href="/dashboard/file/fileA?owner=hello@world.com&selection=sensitive&platformId=gsuite&sensitive=true"
    >
      <div>
        <div>
          Test Filename
        </div>
        <div
          class="FileCell__spacer-top-5"
        >
          <div
            class="sensitiveFileAlert trigger"
            data-testid="sensitiveFileAlert"
          >
            <span
              class="Typography Typography--body Marker Marker__sensitive_content"
            >
              Sensitive Content
            </span>
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;
